src/lib.rs

```diff
@@ -2242,6 +2242,29 @@ extern "C" fn fn_derive_key(
         _ => return CKR_MECHANISM_INVALID,
     };
 
+    #[cfg(feature = "fips")]
+    match mechanism.mechanism {
+        CKM_TLS12_MASTER_KEY_DERIVE
+        | CKM_TLS12_KEY_AND_MAC_DERIVE
+        | CKM_TLS12_KEY_SAFE_DERIVE
+        | CKM_TLS12_KDF
+        | CKM_TLS_KDF => {
```

Comment:
What are the use cases for CKM_TLS12_KDF and CKM_TLS_KDF? We wouldn't consider those mechanisms as approved for FIPS unless there's a TLS protocol usecase that can't be done using MASTER_KEY_DERIVE, KEY_AND_MAC_DERIVE, or KEY_SAFE_DERIVE.

Reply:
Are you saying CKM_TLS[12]_KDF should be made unavailable in FIPS mode? (we can easily do that at build time).

Reply:
SP 800-135r1 approves the KDFs only "in the context of the TLS protocol". So if there is no usecase for CKM_TLS[12]_KDF that is not already covered by the other mechanisms, then it should be non-approved, yes.

Reply:
I think this will be a follow-up change in the indicators waters. For now, we have just HKDF in the fips indicators file so I assume we should also list some of the TLS KDFs mechanisms that we want to allow:
https://github.com/latchset/kryoptic/blob/ff6b251a6ab87956f2eef1440bead7d80032a4e3/src/fips/indicators.rs#L813

Reply:
I am a bit lost in the PKCS#11 specs about what is really needed for the TLS operations. I think you are right and we do not need the `CKM_TLS_KDF`. The specs talks about implementation of `key material exporters`, which does not sounds like something FIPS approved :)
The `CKM_TLS12_KDF` looks like generic algorithm too so lets mark them as unapproved in the indicators and see how it will go. I can do it in this PR once I already touched this. So far we do not have any tls kdf in the fips indicators list.

Reply:
Done also the indicators, including the `CKM_TLS12_MAC`. But I am not exactly sure what sizes/operations limitation should be on all of them, mostly `CKM_TLS12_MAC`. Why the TLS has to be so complicated?

Reply:
This is a bug I brought up for solution in PKCS#11 3.2 specification where we'll deprecate a bunch of duplicated "12" mechanisms that were never needed an never implemented by anybody :(
CKM_TLS_KDF == CKM_TLS12_KDF and CK_TLS_MAC == CKM_TLS12_MAC ...

Reply:
This is an example of the language that there will be in the new spec: "Note: CKM_TLS12_MAC is provided only for historical reasons and should be considered deprecated. This mechanism shares the same behavior with CKM_TLS_MAC."

So we should mark only CKM_TLS_MAC and not CKM_TLS12_MAC as approved as we'll phase out CKM_TLS12_MAC later on once we add a "deprecated" feature.

Reply:
Also this:
```CKM_TLS12_KDF and CKM_TLS12_MAC are considered aliases of the corresponding CKM_TLS_KDF and CKM_TLS_MAC mechanisms and behave identically. These mechanisms are available with a different identifier for historical reasons and are considered deprecated.```

Reply:
Oh, yeah. It looked confusing at places. So the `CKM_TLS_MAC` is the one to use and it only differs from the 1.0/1.1 in the way how the key is derived initially?
In any case, I will update it to use the `CKM_TLS_MAC`. That is tested actually.

Reply:
Yes please

Reply:
I see we use the `CKM_TLS12_MAC` in the `TLS_MASTER_SECRET_ALLOWED_MECHS`. This should be then amended either to support both of them or just the non-deprecated one as the deprecation is not yet public and some people not close to the standardization group might have implemented this.

Reply:
We should accept both mechanism numbers in input but only use the correct variants internally.

Reply:
Seems like the indicators will need some more love too. With current changes, all of the mechanism from tlskdf are considered unapproved, because the test generates the SECRET_KEY of 48b, while the FIPS table has only the keygen approved only for the following size:
```
restrict!(CKK_GENERIC_SECRET, range!(112, 256)),
```
But the keys themselves are approved for larger range:
```
sizes: range!(112, 255 * 64 * 8), 
```
Is there any reason not to allow generic secret keygen of larger sizes?

Reply:
I think GENERIC_SECRET key can be any size